Source/StarRunner2019/HallwayActor.cpp

```diff
@@ -0,0 +1,172 @@
+	{
+		UHallwayUnitComponent *NextHallwayUnitComponent = CreateDefaultSubobject<UHallwayUnitComponent>(FName(*FString::Printf(TEXT("HallwayUnitComponent%d"), i)));
+		FTransform HallwayUnitComponentTransform(Rotator, Translation, Scale);
+
+		NextHallwayUnitComponent->AttachToComponent(GetRootComponent(), FAttachmentTransformRules::KeepRelativeTransform);
+		NextHallwayUnitComponent->SetRelativeTransform(HallwayUnitComponentTransform);
+
+		Translation.X += XLength;
+	}
+
+	HallwayJointComponent = CreateDefaultSubobject<UHallwayJointComponent>(TEXT("HallwayJointComponent"));
+	FTransform HallwayJointComponentTransform(Rotator, Translation, Scale);
+
+	HallwayJointComponent->AttachToComponent(GetRootComponent(), FAttachmentTransformRules::KeepRelativeTransform);
+	HallwayJointComponent->SetRelativeTransform(HallwayJointComponentTransform);
+}
+
+void AHallwayActor::OnOverlapBegin(UPrimitiveComponent* OverlapComponent, AActor* OtherActor, UPrimitiveComponent* OtherComp, int32 OtherBodyIndex, bool bFromSweep, const FHitResult& SweepResult) {
+	if (OtherActor->IsA(AStarRunner2019Character::StaticClass())) {
+		AStarRunner2019Character* Character = Cast<AStarRunner2019Character>(OtherActor);
+		Character->IsTurnable = true;
+		SpawnGrandChildrenHallways();
+	}
+}
+
+void AHallwayActor::SpawnGrandChildrenHallways() {
+	LeftChildHallway->SpawnLeftChildHallway();
+	LeftChildHallway->SpawnRightChildHallway();
+	RightChildHallway->SpawnRightChildHallway();
+	RightChildHallway->SpawnLeftChildHallway();
+}
+
+void AHallwayActor::OnOverlapEnd(UPrimitiveComponent* OverlappedComp, AActor* OtherActor, UPrimitiveComponent* OtherComp, int32 OtherBodyIndex) {
+	if (OtherActor->IsA(AStarRunner2019Character::StaticClass())) {
+		UE_LOG(LogTemp, Warning, TEXT("BEN HUR"));
+		AStarRunner2019Character* Character = Cast<AStarRunner2019Character>(OtherActor);
+		Character->IsTurnable = false;
+
+		if (Character->WentLeft) {
+			DestroyChildHallways(RightChildHallway);
+		}
+		else {
+			DestroyChildHallways(LeftChildHallway);
+		}
+
+		Destroy(); // destroy self
+	}
+}
+
+void AHallwayActor::DestroyChildHallways(AHallwayActor* ChildHallway) {
+	ChildHallway->LeftChildHallway->Destroy();
+	ChildHallway->RightChildHallway->Destroy();
+	ChildHallway->Destroy();
+}
+
+void AHallwayActor::SpawnLeftChildHallway() {
+	UArrowComponent* const LeftChildArrowComponent = Cast<UArrowComponent>(HallwayJointComponent->GetDefaultSubobjectByName(TEXT("LeftArrowComponent")));
+
+	FTransform LeftChildHallwayTransform = GetTransformForComponent(LeftChildArrowComponent);
+	
+	UWorld* const World = GetWorld();
+	if (World) {
+		FActorSpawnParameters Info;
+		Info.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
+
+		LeftChildHallway = World->SpawnActor<AHallwayActor>(AHallwayActor::StaticClass(), LeftChildHallwayTransform, Info);
+	}
+	else {
+		UE_LOG(LogTemp, Warning, TEXT("Nooose"));
+	}
+}
+
+void AHallwayActor::SpawnRightChildHallway() {
```

Comment:
deja vu.
I've just been in this place before
higher on the street, and I know it's my time to go
calling you, and the search is a mystery
standing on my feet, it's so hard to be me, woah

This code is almost the same as SpawnLeftChildHallway.  Maybe move to a helper function?